File: python/Fraction.py
#!/usr/bin/python
from maths import *

class Fraction:
	def __init__(self, numerator, denominator):
		self.numerator = numerator
		self.denominator = denominator
		self.gcd = gcd(self.numerator, self.denominator)
		self.reduced_numerator = int(self.numerator / self.gcd)
		self.reduced_denominator = int(self.denominator / self.gcd)

def addFraction(a, b):
	numerator = a.numerator * b.denominator + b.numerator * a.denominator
	denominator = a.denominator * b.denominator
	result = Fraction(numerator, denominator)
	return result

def multiplyFraction(a, b):
	numerator = a.numerator * b.numerator
	denominator = a.denominator * b.denominator
	result = Fraction(numerator, denominator)
	return result

def divideFraction(a,b):
	numerator =  b.denominator * a.numerator 
	denominator = a.denominator * b.denominator
	result = Fraction(numerator,denominator)
	return result

def substractFraction(a, b):
	numerator = a.numerator * b.denominator - b.numerator * a.denominator
	denominator = a.denominator * b.denominator
	result = Fraction(numerator,denominator)
	return result

def ratioToFraction(a):
	if not isInteger(a.divident) or not isInteger(a.divisor):
		a.divident *= 10
		a.divisor *= 10
		ratioToFraction(a)
	result = Fraction(int(a.divident), int(a.divisor))
	return result

File: c/math/mathFramework/Fraction.c
#include <stdio.h>
#include <stdlib.h>

#include "maths.h"
#include "Fraction.h"

static int reduced_numerator (struct Fraction *this) {
	return this->numerator / gcd(this->numerator, this->denominator);
}

static int reduced_denominator (struct Fraction *this) {
	return this->denominator / gcd(this->numerator, this->denominator);
}

static int fractGcd (struct Fraction *this) {
	return gcd(this->numerator, this->denominator);
}

static struct Fraction new (int numerator, int denominator) {
	return (struct Fraction) {
		.numerator = numerator, .denominator = denominator, .fractGcd = &fractGcd, .reduced_numerator = &reduced_numerator, .reduced_denominator = &reduced_denominator
	};
}

const struct FractionClass Fraction = {
	.new = &new
};

fract addFraction (struct Fraction a, struct Fraction b) {
	return Fraction.new(a.numerator * b.denominator + b.numerator * a.denominator, a.denominator * b.denominator);
}

fract subtractFraction (struct Fraction a, struct Fraction b) {
	return Fraction.new(a.numerator * b.denominator - b.numerator * a.denominator, a.denominator * b.denominator);
}

fract multiplyFraction(struct Fraction a, struct Fraction b) {
	return Fraction.new(a.numerator * b.numerator, a.denominator * b.denominator);
}

fract divideFraction(struct Fraction a, struct Fraction b) {
	return Fraction.new(a.numerator * b.denominator, a.denominator * b.denominator);
}
